src/Mutation.tsx

```diff
@@ -0,0 +1,237 @@
+  mutation: DocumentNode;
+  variables?: TVariables;
+  optimisticResponse?: Object;
+  refetchQueries?: string[] | PureQueryOptions[];
+  update?: MutationUpdaterFn;
+  children: (
+    mutateFn: () => void,
+    result?: MutationResult<TData>,
+  ) => React.ReactNode;
+  onCompleted?: (data: TData) => void;
+  onError?: (error: ApolloError) => void;
+}
+
+export interface MutationState<TData = any> {
+  notCalled: boolean;
+  error?: ApolloError;
+  data?: TData;
+  loading?: boolean;
+}
+
+const initialState = {
+  notCalled: true,
+};
+
+class Mutation<
+  TData = any,
+  TVariables = OperationVariables
+> extends React.Component<
+  MutationProps<TData, TVariables>,
+  MutationState<TData>
+> {
+  static contextTypes = {
+    client: PropTypes.object.isRequired,
+  };
+  
+  static propTypes = {
+    mutation: PropTypes.object.isRequired,
+    variables: PropTypes.object,
+    optimisticResponse: PropTypes.object,
+    refetchQueries: PropTypes.oneOfType([
+      PropTypes.arrayOf(PropTypes.string),
+      PropTypes.arrayOf(PropTypes.object)
+    ]),
+    update: PropTypes.func,
+    children: PropTypes.func.isRequired,
+    onCompleted: PropTypes.func,
+    onError: PropTypes.func
+  };
+
+  private client: ApolloClient<any>;
+  private mostRecentMutationId: number;
+  
+  constructor(props: MutationProps<TData, TVariables>, context: any) {
+    super(props, context);
+
+    this.verifyContext(context);
+    this.client = context.client;
+
+    this.verifyDocumentIsMutation(props.mutation);
+    
+    this.mostRecentMutationId = 0;
+    this.state = initialState;
+  }
+
+  componentWillReceiveProps(nextProps, nextContext) {
+    if (
+      shallowEqual(this.props, nextProps) &&
+      this.client === nextContext.client
+    ) {
+      return;
+    }
+
+    if (this.props.mutation !== nextProps.mutation) {
+      this.verifyDocumentIsMutation(nextProps.mutation);
+    }
+
+    if (this.client !== nextContext.client) {
+      this.client = nextContext.client;
+      this.setState(initialState);
+    }
+  }
+
+  render() {
+    const { children } = this.props;
+    const { loading, data, error, notCalled } = this.state;
+
+    const result = notCalled
+      ? undefined
+      : {
+          loading,
+          data,
+          error,
+        };
+
+    return children(this.runMutation, result);
+  }
+
+  private runMutation = async () => {
+    this.onStartMutation();
+    
+    this.mostRecentMutationId = this.mostRecentMutationId + 1;
+    const mutationId = this.mostRecentMutationId;
+    
+    try {
+      const response = await this.mutate();
+      this.onCompletedMutation(response, mutationId);
+    } catch (e) {
+      this.onMutationError(e, mutationId);
+    }
+  };
+
+  private mutate = async () => {
+    
+    const {
+      mutation,
+      variables,
+      optimisticResponse,
+      refetchQueries,
+      update,
+    } = this.props;
+
+    const response = await this.client.mutate({
```

Comment:
why not return the await'd mutate function instead of creating a variable?

Reply:
Good point! I updated the PR:) Thanks